ProcFilter.cpp

```diff
@@ -0,0 +1,307 @@
+}
+
+bool ProcFilter::ParseConfig(const Config& config) {
+    if (config.HasKey(CONFIG_PARAM_NAME)) {
+        auto doc = config.GetJSON(CONFIG_PARAM_NAME);
+        if (!doc.IsObject()) {
+            return false;
+        }
+        for (auto it = doc.MemberBegin(); it != doc.MemberEnd(); ++it) {
+            if (it->value.IsArray()) {
+                for (auto it2 = it->value.Begin(); it2 != it->value.End(); ++it2) {
+                    if (it2->IsString()) {
+                        _filters.insert(std::pair<std::string, std::string>(std::string(it->name.GetString(), it->name.GetStringLength()),
+                                                                            std::string(it2->GetString(), it2->GetStringLength())));
+                    } else {
+                        Logger::Error("Invalid entry (%s) in config for '%s'", it->name.GetString(), CONFIG_PARAM_NAME.c_str());
+                        _filters.clear();
+                        return false;
+                    }
+                }
+            } else {
+                Logger::Error("Invalid entry (%s) in config for '%s'", it->name.GetString(), CONFIG_PARAM_NAME.c_str());
+                _filters.clear();
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+void ProcFilter::Load()
+{
+    if (_filters.empty()) {
+        return;
+    }
+    _previous_filter_pids.clear();
+    _previous_filter_pids = _filter_pids;
+    gettimeofday(&_last_time_initiated, NULL);
+    _filter_pids.clear();
+    // scan existing processes and choose those in the names list and children
+    std::list<ProcessInfo>* listOfProcesses = get_all_processes();
+    compile_filter_pids(listOfProcesses);
+    delete listOfProcesses;
+}
+
+ProcFilter::ProcFilter(const std::shared_ptr<UserDB>& user_db)
+{
+    _user_db = user_db;
+}
+
+bool ProcFilter::ShouldFilter(int pid, int ppid)
+{
+    return (_filter_pids.find(pid) != _filter_pids.end() || _filter_pids.find(ppid) != _filter_pids.end() || _previous_filter_pids.find(pid) != _previous_filter_pids.end() || _previous_filter_pids.find(ppid) != _previous_filter_pids.end());
+}
+
+bool ProcFilter::test_and_recompile()
+{
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    if(tv.tv_sec - _last_time_initiated.tv_sec > RELOAD_INTERVAL)
+    {
+        Load();
+        return true;
+    }
+    return false;
+}
+
+void ProcFilter::AddProcess(int pid, int ppid)
+{
+
+    // Do nothing if there are no filters
+    if (_filters.empty()) 
+    {
+        return;
+    }  
+    
+    // Check to see if the entire set needs to be re-initialized
+    if(test_and_recompile())
+    {
+        return;
+    }
+
+    // This new processes's pid might still be present in the list if the pid was used by a previous process.
+    // So, remove it from the list. If this new process needs to be filtered it will get re-added during the
+    // parent pid check, or during the root filter prod check.
+    _filter_pids.erase(pid);
```

Comment:
Should also delete pid from _previous_filter_pids.

Reply:
the previous one is a history. That's all the point not to delete from there so the already dead processes are blocked